src/System.Threading/tests/ReaderWriterLockSlimTests.cs

```diff
@@ -380,5 +381,59 @@ public static void WriterToUpgradeableReaderChain()
             }
         }
 
+        [Fact]
+        public static void ReleaseReadersWhenWaitingWriterTimesOut()
+        {
+            using (var rwls = new ReaderWriterLockSlim())
+            {
+                // Enter the read lock
+                rwls.EnterReadLock();
+                // Typical order of execution: 0
+
+                Task writeWaiterTask;
+                using (var beforeTryEnterWriteLock = new ManualResetEvent(false))
+                {
+                    writeWaiterTask =
+                        Task.Run(() =>
+                        {
+                            // Typical order of execution: 1
+
+                            // Add a writer to the wait list for enough time to allow successive readers to enter the wait list while this
+                            // writer is waiting
+                            beforeTryEnterWriteLock.Set();
+                            if (rwls.TryEnterWriteLock(2000))
+                            {
+                                // The typical order of execution is not guaranteed, as sleep times are not guaranteed. For
+                                // instance, before this write lock is added to the wait list, the two new read locks may be
+                                // acquired. In that case, the test may complete before or while the write lock is taken.
+                                rwls.ExitWriteLock();
+                            }
+
+                            // Typical order of execution: 4
+                        });
+                    beforeTryEnterWriteLock.WaitOne();
+                }
+                Thread.Sleep(1000); // wait for TryEnterWriteLock to enter the wait list
+
+                // A writer should now be waiting, add readers to the wait list. Since a read lock is still acquired, the writer
+                // should time out waiting, then these readers should enter and exit the lock.
+                Action<int> EnterReadLockNumber =
+                    (int readLockNumber) =>
+                    {
+                        // Typical order of execution: 2, 3
+                        rwls.EnterReadLock();
+                        // Typical order of execution: 5, 6
+                        rwls.ExitReadLock();
+                    };
+                Task.WaitAll(
```

Comment:
This may explain the failure.  If one of these becomes a recursive attempt, then it will throw (since you disabled recursion).  That will invoke the `using` statement's `finally`, disposing the RWLS while the read lock is still held, resulting in a second exception (which is the one we see in the test results).